www/src/public/js/controllers/builder/main.js

```diff
@@ -2070,6 +2308,332 @@
             return total;
         };
 
+        /** Update runecraft charms. */
+        $scope.runeCraftCharms = function($event) {
+            console.log(event);
+
+            var cslot = $scope.currentItemIndex;
```

Comment:
There appears to be a lot of duplicate code in this function, which can be difficult to maintain. It also creates more opportunities for code to be wrong. The really difficult thing about duplicate code like this, is that if there is a bug in one of these `case` statements, but not any of the others, it's much harder to find because it only happens some times.

There's a really cool trick in JavaScript that you can use here to help reduce a bunch of this code.
In most languages, you're probably used to accessing variables in an object via the following syntax: `$scope.selectedList.runeCharms.charm1`. In JavaScript, however, you can also access them via string key by using the indexing operators: `$scope["selectedList"]["runeCharms"]["charm1"]`. Now, most of the time, you're going to want to use the former syntax, because it's clearer to anyone reading it, but it can be really useful to index into the object instead, especially in this instance. I'm not going to write out all of the code you need, but hopefully this will point you in the right direction:

```
var cslot = $scope.currentItemIndex;
var charmSlot = undefined;
switch (cslot) {
  case 3:
    charmSlot = "charm1";
    break;
  case 4:
    charmSlot = "charm2";
    break;
  case 14:
    charmSlot = "charm3";
    break;
  case 15:
    charmSlot = "charm4";
    break;
  default:
    break;
}

if (charmSlot) {
  $scope.selectedList.runeCharms[charmSlot] = $scope.charm1Selector; // note that this is = while the rest are +=
  $scope.selectedList.runeCharms[charmSlot] += $scope.charm2Selector;
  $scope.selectedList.runeCharms[charmSlot] += $scope.charm3Selector;
  $scope.selectedList.runeCharms[charmSlot] += $scope.charm4Selector;
  $scope.selectedList.runeCharms[charmSlot] += $scope.charm5Selector;

  // ... other rune charm logic
}
```

Now you shouldn't need to put any other logic inside that switch/case statement, just the charmSlot logic.

Reply:
nice. I was wondering what you would come up with for this.

Reply:
much cleaner!